src/logic/taskcluster.ts

```diff
@@ -152,3 +158,143 @@ export async function retrieveTaskclusterUserCredentials(
 
   return response.json() as Promise<UserCredentials>;
 }
+
+type Action = {
+  hookId: string;
+  hookGroupId: string; // check to see if it used
+  kind: 'hook';
+  hookPayload: unknown;
+  name: string;
+};
+
+export async function fetchActionsFromDecisionTask(
+  rootUrl: string,
+  decisionTaskId: string,
+) {
+  const url = `${rootUrl}/api/queue/v1/task/${decisionTaskId}/artifacts/public%2Factions.json`;
+
+  const response = await fetch(url);
+
+  void checkTaskclusterResponse(response);
+
+  return (await response.json()) as {
+    actions: Array<Action>;
+    variables: unknown;
+  };
+}
+
+export async function fetchTaskInformationFromTaskId(
+  rootUrl: string,
+  taskId: string,
+) {
+  const url = `${rootUrl}/api/queue/v1/task/${taskId}`;
+
+  const response = await fetch(url);
+
+  void checkTaskclusterResponse(response);
+
+  return (await response.json()) as { scopes: Array<string> };
+}
+
+export async function expandScopes(rootUrl: string, scopes: Array<string>) {
+  const url = `${rootUrl}/api/auth/v1/scopes/expand`;
+
+  const options = {
+    method: 'POST',
+    body: JSON.stringify({ scopes }),
+    headers: { 'Content-Type': 'application/json' },
+  };
+
+  const response = await fetch(url, options);
+
+  void checkTaskclusterResponse(response);
+
+  const json = (await response.json()) as { scopes: Array<string> };
+
+  return json.scopes;
+}
+
+export async function triggerHook(
+  rootUrl: string,
+  accessToken: string,
+  hookGroupId: string,
+  hookId: string,
+  hookPayload: string,
+) {
+  const url = `${rootUrl}/api/hooks/v1/hooks/${encodeURIComponent(
+    hookGroupId,
+  )}/${encodeURIComponent(hookId)}/trigger`;
+
+  const options = {
+    method: 'POST',
+    body: JSON.stringify(hookPayload),
+    headers: {
+      Authorization: `Bearer ${accessToken}`, // to be checked
+      'Content-Type': 'application/json',
+    },
+  };
+
+  const response = await fetch(url, options);
+
+  void checkTaskclusterResponse(response);
+
+  const json = (await response.json()) as { taskId: string };
+
+  return json.taskId;
+}
+
+export async function retrigger(rootUrl: string, repo: string, jobId: number) {
+  // GOAL: retrigger one job
+
+  const jobInfo = await fetchJobInformationFromJobId(repo, jobId);
+
+  const { push_id: pushId } = jobInfo;
+  const decisionTaskId = await fetchDecisionTaskIdFromPushId(repo, pushId);
+
+  const actionsResponse = await fetchActionsFromDecisionTask(
+    rootUrl,
+    decisionTaskId,
+  );
+
+  const retriggerAction = actionsResponse.actions.find(
+    (action) => action.name === 'retrigger-multiple',
+  );
+
+  // submit retrigger action to Taskcluster
+  const context = Object.assign(
+    {},
+    {
+      taskGroupId: decisionTaskId,
+      taskId: null,
+      input: jobInfo.job_type_name,
+    },
+    actionsResponse.variables,
+  );
+
+  if (retriggerAction?.kind === 'hook') {
+    // maybe use JSON.stringify
```

Comment:
why not, but in that case remove it from `triggerHook`.

Reply:
Actually triggerHook's signature is currently incorrect, hookPayload is a string but it's really an object now. Calling JSON.stringify here (or rather in the call below) would make it correct.